platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbCompactionCoordinator.java

```diff
@@ -0,0 +1,280 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.merkledb;
+
+import static com.swirlds.common.threading.manager.AdHocThreadManager.getStaticThreadManager;
+import static com.swirlds.logging.LogMarker.EXCEPTION;
+import static com.swirlds.logging.LogMarker.MERKLE_DB;
+import static com.swirlds.merkledb.MerkleDb.MERKLEDB_COMPONENT;
+import static java.util.Objects.requireNonNull;
+
+import com.swirlds.common.config.singleton.ConfigurationHolder;
+import com.swirlds.common.threading.framework.config.ThreadConfiguration;
+import com.swirlds.merkledb.config.MerkleDbConfig;
+import com.swirlds.merkledb.files.DataFileCompactor;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.io.IOException;
+import java.nio.channels.ClosedByInterruptException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This class is responsible for coordinating compaction tasks for a {@link MerkleDbDataSource}.
+ * It provides convenient API for starting compactions for each of the three storage types. Also, this class makes sure
+ * that there are no concurrent compactions for the same storage type. And finally it provides a way to stop all compactions
+ * and keep them disabled until they are explicitly enabled again.
+ * The compaction tasks are executed in a background thread pool.
+ * The number of threads in the pool is defined by {@link MerkleDbConfig#compactionThreads()} property.
+ *
+ */
+class MerkleDbCompactionCoordinator {
+
+    /**
+     * Since {@code com.swirlds.platform.Browser} populates settings, and it is loaded before any
+     * application classes that might instantiate a data source, the {@link ConfigurationHolder}
+     * holder will have been configured by the time this static initializer runs.
+     */
+    private static final MerkleDbConfig config = ConfigurationHolder.getConfigData(MerkleDbConfig.class);
+
+    /**
+     * A thread pool to run compaction tasks.
+     */
+    private static final AtomicReference<ExecutorService> compactionExecutorServiceRef = new AtomicReference<>();
```

Comment:
As far as I can tell, this executor is initialized in constructor and never changed afterwards. Why should be be wrapped into AtomicReference then? Can it be just a final field? Declaring something final is the easiest way to make it thread safe

Reply:
Yes, you're right. The reference made sense when the initializing method recreated the executor if it was previously shut down. Now the initialization could be simplified to the static block.